src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
-        var envelope = self.allocator.create(Envelope) catch unreachable;
+        const envelope = self.allocator.create(Envelope) catch unreachable;
         envelope.* = .{
-            .address = .{ .replica = self.configuration[replica].replica },
+            .address = self.configuration[replica],
             .message = message,
-            .next = null,
         };
-        self.ready.push(envelope);
+
+        // Messages sent by the server to itself are delivered directly in flush()
+        if (replica == self.server.replica) {
+            self.replicas[replica].send_queue.push(envelope);
+        } else {
+            self.replicas[replica].send_message(envelope);
+        }
     }
 
-    pub fn send_queued_messages(self: *MessageBus) void {
-        while (self.ready.out != null) {
-            // Loop on a copy of the linked list, having reset the linked list first, so that any
-            // synchronous append in on_message() is executed the next iteration of the outer loop.
-            // This is not critical for safety here, but see run() in src/io.zig where it is.
-            var copy = self.ready;
-            self.ready = .{};
-            while (copy.pop()) |envelope| {
-                assert(envelope.message.references > 0);
-                switch (envelope.address) {
-                    .replica => |r| r.on_message(envelope.message),
-                }
-                envelope.message.references -= 1;
-                self.gc(envelope.message);
-                // The lifetime of an envelope will often be shorter than that of the message.
-                self.allocator.destroy(envelope);
+    pub fn send_header_to_client(self: *MessageBus, client_id: u128, header: Header) void {
+        assert(header.size == @sizeOf(Header));
+
+        // TODO Pre-allocate messages at startup.
+        var message = self.create_message(@sizeOf(Header)) catch unreachable;
+        message.header.* = header;
+
+        const data = message.buffer[@sizeOf(Header)..message.header.size];
+        // The order matters here because checksum depends on checksum_data:
+        message.header.set_checksum_data(data);
+        message.header.set_checksum();
+
+        assert(message.references == 0);
+        self.send_message_to_client(replica, message);
+    }
+
+    /// Try to send the message to the client with the given id.
+    /// If the client is not currently connected, the message is silently dropped.
+    pub fn send_message_to_client(self: *MessageBus, client_id: u128, message: *Message) void {
+        for (self.connections) |connection| {
+            switch (connection) {
+                .client => |id| if (id == client_id) {
+                    message.references += 1;
+                    // TODO Pre-allocate envelopes at startup.
+                    const envelope = self.allocator.create(Envelope) catch unreachable;
+                    envelope.* = .{
+                        .address = self.configuration[replica],
```

Comment:
Why do we use `self.configuration[replica]` as the destination `address` here when sending to the client? Should we not just use undefined? Do we still need these envelope addresses or even envelopes at all now?

Reply:
I used them in the first MessageBus shim, just to know where to deliver to when processing the queue, but we've switched to using a queue per replica here, instead of a single queue, and your change is much better, even when we shim again for testing.

Reply:
I actually already removed the address field from `Envelope` but forgot to remove the initialization here.

We do still need envelopes as I believe a single `Message` might be sent to multiple recipients at the same time, and because `vr.zig` uses `Message.next` for its own purposes.

Reply:
Ah yes, while we could add another `next` to `Message` to solve the `vr.zig` repair queue usage, as you say that wouldn't work for when the message needs to be in multiple send queues at the same time.